extensions/pkg/controller/worker/machines.go

```diff
@@ -120,6 +120,11 @@ func (m MachineDeployments) HasSecret(secretName string) bool {
 
 // WorkerPoolHash returns a hash value for a given worker pool and a given cluster resource.
 func WorkerPoolHash(pool extensionsv1alpha1.WorkerPool, cluster *extensionscontroller.Cluster, additionalData ...string) (string, error) {
+	return WorkerPoolHashWithProviderConfigOption(pool, cluster, true, additionalData...)
+}
+
+// WorkerPoolHashWithProviderConfigOption returns a hash value for a given worker pool and a given cluster resource.
+func WorkerPoolHashWithProviderConfigOption(pool extensionsv1alpha1.WorkerPool, cluster *extensionscontroller.Cluster, includeProviderConfig bool, additionalData ...string) (string, error) {
```

Comment:
Instead of adding a new function, what about enhancing the signature of `WorkerPoolHash` with this new `includeProviderConfig bool` argument?

Reply:
> Instead of adding a new function, what about enhancing the signature of WorkerPoolHash with this new includeProviderConfig bool argument?

I tried to avoid making this a breaking change by providing the new function and implementing the old behavior on top. We can of course deprecate the old function directly and skip the middle step.

> Also, can you explain how the process looks like when a provider extension wants to migrate to the new WorkerPoolHashWithProviderConfigOption function in a way to prevent existing machines from being rolled out?

There is no easy way to migrate unfortunately. At this moment, just adding anything in the providerConfig (`PC`) will trigger the change in hash even if  there is no "actual" change that needs to happen to the workers. Therefore, at some point you will have to "eat the frog" migrate to the new function - triggering the rollout in the process- but regaining control over the hash.

The only realistic alternative currently is a variation of what you see [here](https://github.com/gardener/gardener-extension-provider-openstack/blob/master/pkg/controller/worker/machines.go#L257-L265) -  basically try to apply transformations to a PC to discard fields that should not be used as part of the hash but I am not sure how well this can look like in the library. This may help you avoid some cases that you would rollout but its more complicated in the long run.
 
 cc: @timuthy

Reply:
> I tried to avoid making this a breaking change by providing the new function and implementing the old behavior on top. We can of course deprecate the old function directly and skip the middle step.

Personally, to reduce some effort, I would indeed skip the middle step and directly change the signature. Extensions depending on this need to simply adapt when they revendor a new version of g/g which anyways is a manual/custom/human activity.

> Therefore, at some point you will have to "eat the frog" migrate to the new function - triggering the rollout in the process- but regaining control over the hash.

Can you explain more about this? Wouldn't the easiest approach be to combine this with a Kubernetes version upgrade? I.e., starting from Kubernetes 1.27, the worker pool hash function shall no longer consider the entire provider config but only the selectively added fields?

Reply:
> Can you explain more about this? Wouldn't the easiest approach be to combine this with a Kubernetes version upgrade? I.e., starting from Kubernetes 1.27, the worker pool hash function shall no longer consider the entire provider config but only the selectively added fields?

The issue with tying this change with a k8s update is that wouldn't be able to implement any "non-rolling" change to the worker config until v1.27. Or alternatively, any of the workerConfigSpec changes would only become non-rolling when you upgrade to the new k8s version. Frankly we haven't got that many, otherwise the need for this change would have become apparent sooner - so we can take either route for now.

Also simply switching the `includeProviderConfig` boolean flag based on the shoot k8s version implicitly could also represent some danger in the sense during the rebasing to the gardener version, you would need to write a function to properly calculate your `additionalfields` at the time of the revendoring. However,  this is not a major issue for our repos either.